src/utils/helpers/revalidateAll.tsx

```diff
@@ -0,0 +1,73 @@
+import {
+  ApplicationElementStates,
+  ResponsesByCode,
+  RevalidateResult,
+  ValidityFailure,
+} from '../types'
+import { defaultValidate } from '../../formElementPlugins/defaultValidate'
+
+export const revalidateAll = async (
+  elementsState: ApplicationElementStates | undefined,
+  responsesByCode: ResponsesByCode | undefined,
+  strict = true,
+  shouldUpdateDatabase = true
+): Promise<RevalidateResult> => {
+  const validate = defaultValidate // To-Do: import custom validation methods
+
+  // console.log('Elements', elementsState)
+  // console.log('responses', responsesByCode)
+  if (elementsState) {
+    const elementCodes = Object.keys(elementsState).filter(
+      (key) =>
+        responsesByCode && // Typescript requires this
+        elementsState[key].category === 'QUESTION' &&
+        elementsState[key].isVisible === true &&
+        // Strict/Loose validation logic:
+        ((strict && elementsState[key].isRequired) ||
+          (strict && responsesByCode[key]?.isValid !== null) ||
+          (!strict && responsesByCode[key]?.isValid !== null))
+    )
+
+    const validationExpressions = elementCodes.map((code) => elementsState[code].validation)
+
+    const evaluatedValidations = []
+
+    for (let i = 0; i < elementCodes.length; i++) {
+      const code = elementCodes[i]
```

Comment:
I think this would have been more elegant using instead an array that keeps both: elementCode and validation expression, so the first `. map` can be changes to: `const validationExpressions = elementCodes.map((code) => ({ code, expression: elementsState[code].validation }))` then you iterate over this array using both properties: `validationExpressions.map(({code, expression}) => ... )`

Reply:
Yeah, I agree that this would be nicer. The reason I didn't is that the `Promise.all` requires an array of promises, and it won't work if the promise is only a field within an object in the array.

Does that make sense, or is the a simple way around this that you can think of?

Reply:
As far as I'm understanding the code you get as array of promises with `evaluatedValidations` which is created inside this loop. What I'm saying is that this loop can be run using the .map function instead of the traditional for loop

Reply:
Ah -- I think I get you now. Good suggestion, have updated it, seems cleaner. Thanks :) fabc757